geomesa-spark-sql/geomesa-spark-sql-core/src/main/scala/org/apache/spark/sql/SQLSpatialFunctions.scala

```diff
@@ -0,0 +1,113 @@
+  val ST_Contains:   (Geometry, Geometry) => Boolean = (geom1, geom2) => geom1.contains(geom2)
+  val ST_Covers:     (Geometry, Geometry) => Boolean = (geom1, geom2) => geom1.covers(geom2)
+  val ST_Crosses:    (Geometry, Geometry) => Boolean = (geom1, geom2) => geom1.crosses(geom2)
+  val ST_Disjoint:   (Geometry, Geometry) => Boolean = (geom1, geom2) => geom1.disjoint(geom2)
+  val ST_Equals:     (Geometry, Geometry) => Boolean = (geom1, geom2) => geom1.equals(geom2)
+  val ST_Intersects: (Geometry, Geometry) => Boolean = (geom1, geom2) => geom1.intersects(geom2)
+  val ST_Overlaps:   (Geometry, Geometry) => Boolean = (geom1, geom2) => geom1.overlaps(geom2)
+  val ST_Touches:    (Geometry, Geometry) => Boolean = (geom1, geom2) => geom1.touches(geom2)
+  val ST_Within:     (Geometry, Geometry) => Boolean = (geom1, geom2) => geom1.within(geom2)
+
+  val ST_Centroid: Geometry => Point = g => g.getCentroid
+  val ST_DistanceSpheroid: (Geometry, Geometry) => java.lang.Double = (s, e) => fastDistance(s, e)
+
+  // Geometry Processing
+  val ch = new ConvexHull
+
+  // Type casting functions
+  // TODO: Implement addition casts
+  val ST_CastToPoint:      Geometry => Point       = g => g.asInstanceOf[Point]
+  val ST_CastToPolygon:    Geometry => Polygon     = g => g.asInstanceOf[Polygon]
+  val ST_CastToLineString: Geometry => LineString  = g => g.asInstanceOf[LineString]
+
+  def registerFunctions(sqlContext: SQLContext): Unit = {
+    // Register geometry constructors
+    sqlContext.udf.register("st_geomFromWKT"   , ST_GeomFromWKT)
+    sqlContext.udf.register("st_makeBox2D"     , ST_MakeBox2D)
+    sqlContext.udf.register("st_makeBBOX"      , ST_MakeBBOX)
+
+    // Register geometry accessors
+    sqlContext.udf.register("st_envelope"      , ST_Envelope)
+
+    // Register geometry editors
+    sqlContext.udf.register("st_translate", ST_Translate)
+
+    // Register spatial relationships
+    sqlContext.udf.register("st_contains"    , ST_Contains)
+    sqlContext.udf.register("st_covers"      , ST_Covers)
+    sqlContext.udf.register("st_crosses"     , ST_Crosses)
+    sqlContext.udf.register("st_disjoint"    , ST_Disjoint)
+    sqlContext.udf.register("st_equals"      , ST_Equals)
+    sqlContext.udf.register("st_intersects"  , ST_Intersects)
+    sqlContext.udf.register("st_overlaps"    , ST_Overlaps)
+    sqlContext.udf.register("st_touches"     , ST_Touches)
+    sqlContext.udf.register("st_within"      , ST_Within)
+
+    sqlContext.udf.register("st_centroid"      , ST_Centroid)
+    sqlContext.udf.register("st_distanceSpheroid"  , ST_DistanceSpheroid)
+
+    // Register geometry Processing
+    sqlContext.udf.register("st_convexhull", ch)
+
+    // Register type casting functions
+    sqlContext.udf.register("st_castToPoint", ST_CastToPoint)
+  }
+
+  @transient private val geoCalcs = new ThreadLocal[GeodeticCalculator] {
+    override def initialValue(): GeodeticCalculator = new GeodeticCalculator(DefaultGeographicCRS.WGS84)
```

Comment:
can use `org.locationtech.geomesa.utils.geotools.CRS_EPSG_4326` which will use the full CRS if available - not sure if it matters...